### DISCUSSION TITLE
Source generators - how to generate non source files?

### DISCUSSION THREAD

**Discussion by torgabor** [dotnet/roslyn#54566] [Q&A]
Hi! 
I've started using Source Generators, and so far I've been having a great time. I'm generating serialization code for an in-house protocol format. I'd also like to generate a protocol descriptor JSON file and include it into the project as a file as well.  Is there a way to do this currently?

The workaround I found was to use  set a  `CompilerVisibleProperty` to `MSBuildProjectDirectory` and manually write the file, however this seems to me to be a quite hacky solution.

**Comment by RikkiGibson** [comment]
Maybe #54798 solves your problem?

**Reply by torgabor** [reply]
Thanks for the answer - however looking at the proposed - and now merged pull request, but as far as I could tell, it only supports non-semantic sources - meaning C# code that  needs to be compiled, but shouldn't be directly user-visible.

The intention does seem to be there, as referenced by 
[https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md](url)
However, the API is subtly different from the design, and the parts related to generating non-source artifacts are missing.

**Reply by RikkiGibson** [reply]
@chsienki is it possible for a source generator to generate non-source files? i.e. files that shouldn't be added to the compilation at all but simply written out to disk?

**Reply by chsienki** [reply]
Not currently, but it's been a highly requested feature. We've already started work on enabling this for analyzers (https://github.com/dotnet/roslyn/pull/49046) and have a design in the works for generators. 

I don't want to commit to timelines, but my expectation is that we'll enable this scenario in the shorter rather than longer term.
